[MM-38770] Show a copy button when hovering over code blocks

cleferman opened the pull request:
<!-- Thank you for contributing a pull request! Here are a few tips to help you:

1. If this is your first contribution, make sure you've read the Contribution Checklist https://developers.mattermost.com/contribute/getting-started/contribution-checklist/
2. Read our blog post about "Submitting Great PRs" https://developers.mattermost.com/blog/2019-01-24-submitting-great-prs
3. Take a look at other repository specific documentation at https://developers.mattermost.com/contribute
-->

#### Summary
<!--
A description of what this pull request does.
-->

A copy button is now shown when hovering over a text block. Upon clicking it the code block gets copied into the clipboard and transitions into a checkmark to show copying the text was a success.

#### Ticket Link
<!--
If this pull request addresses a Help Wanted ticket, please link the relevant GitHub issue, e.g.

  Fixes https://github.com/mattermost/mattermost-server/issues/XXXXX

Otherwise, link the JIRA ticket.
-->
Jira ticket: https://mattermost.atlassian.net/browse/MM-38770
Fixes https://github.com/mattermost/mattermost-server/issues/18851

#### Related Pull Requests
<!--
List all PRs related to resolving a ticket. For instance, if you submitted a PR to `mattermost/mattermost-server`, please include it here.
-->
- Has server changes (please link here)
- Has mobile changes (please link here)

#### Screenshots
<!--
If the PR includes UI changes, include screenshots/GIFs.

For an easier comparison of UI changes a table (template below) can be used.

|  before  |  after  |
|----|----|
| <insert before screenshot here> | <insert after screenshot here> |
![image](https://user-images.githubusercontent.com/14358537/141173112-23254f61-bcba-4701-be40-5d90480089f4.png)

-->

#### Release Note
<!--
Add a release note for each of the following conditions:

* Config changes (additions, deletions, updates)
* API additions—new endpoint, new response fields, or newly accepted request parameters
* Database changes (any)
* Websocket additions or changes
* Anything noteworthy to a Mattermost instance administrator (err on the side of over-communicating)
* New features and improvements, including behavioural changes, UI changes and CLI changes
* Bug fixes and fixes of previous known issues
* Deprecation warnings, breaking changes, or compatibility notes

If no release notes are required write NONE. Use past-tense. Newlines are stripped.

Examples:

```
Added new API endpoints POST /api/v4/foo, GET api/v4/foo, and GET api/v4/foo/:foo_id.
```

```
Added a new config setting ServiceSettings.FooBar. Added a new column Foo to the Users table.
```

```
NONE
```
-->
```release-note
Added copy button when hovering over code blocks
```

mattermod commented:
Hello @cleferman,

Thanks for your pull request! A Core Committer will review your pull request soon. For code contributions, you can learn more about the review process [here](https://developers.mattermost.com/contribute/getting-started/code-review/).

hmhealey commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/106352

jgilliam17 commented:
/update-branch

hmhealey commented:
Removing the Dev Review label since I forgot that we had a second review on the original PR

matthewbirtch commented:
> copy button + label background on dark themes appears off. @matthewbirtch would you be able to provide a dark theme example of expected in figma? Thanks

@jgilliam17 this may actually be an issue with the `Indigo` theme. For some reason it's showing a white background for code blocks. Could you file a ticket to fix this?

This CSS seems to be setting the background and text colors for code
![image](https://user-images.githubusercontent.com/2040554/151234602-294e2603-0716-4af0-8903-6abfa798f2af.png)

jgilliam17 commented:
@matthewbirtch  sounds good, I'll add a ticket for that.
Should we center the label vertically ? I feel it needs to be moved up few pixels 
<img width="322" alt="Screen Shot 2022-01-26 at 2 43 42 PM" src="https://user-images.githubusercontent.com/52937121/151235367-c5b88c6e-711a-4043-9fb5-afc6afd1473c.png">

matthewbirtch commented:
> Should we center the label vertically ? I feel it needs to be moved up few pixels

The background color behind the label and copy icon won't be distinguishable because the code background should be the same color - so you won't notice any off-centering. We could/should add bottom padding to that container though so it does sit vertically centered, but I think it was removed because it made it too high for single-line code

jgilliam17 commented:
Thanks Matt
-  Ticket added for Indigo theme codeblock background https://mattermost.atlassian.net/browse/MM-41299

jgilliam17 commented:
@cleferman 
One test is failing on Cypress, MM-T678 (see full report[ here](https://mattermost-cypress-report.s3.amazonaws.com/106352-57dcc47-onprem-ent-webapp-pr-9655-mattermost/mm-ee-test57dcc47/mochawesome.html))
posting this text, causes the white screen and console errors 
`/code 1. Not a list item, **not bolded**, http://notalink.com, ~off-topic is not a link to the channel.`
<img width="1265" alt="Screen Shot 2022-01-26 at 3 10 26 PM" src="https://user-images.githubusercontent.com/52937121/151239193-28775df6-313d-407e-87eb-4eb26ff2216e.png">

cleferman commented:
@jgilliam17 I'll have a look next week

jgilliam17 commented:
/update-branch

cleferman commented:
@jgilliam17 sorry it took so long. Please check :).
I'll be afk the entire week. If any other issues arise and it's needed urgently feel free to assign it to someone else until I get back. 
Cheers

matthewbirtch commented:
@cleferman, I just had a quick look and it looks like the lines of code no longer align with the line numbers on the left
![image](https://user-images.githubusercontent.com/2040554/152906274-2211e89d-f96f-4410-9cf1-dd4d7da83877.png)

jgilliam17 commented:
/e2e-test

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/148302

cleferman commented:
@jgilliam17 I updated the HTML assertions for failing markdown tests. They were failing because the DOM changed to accommodate the copy button and the context menu. The `channel files search` test is unrelated to this feature. However, it seems to be passing right now (on my machine) without doing anything :).

matthewbirtch reviewed (changes requested):
@cleferman looks like the code lines and the numbers are aligning better now. Thanks!

What can we do to make the background of this container always the same color as the code area background? I can't seem to find the right variable for this.

<img width="240" alt="153886711-4eb9fe58-208a-4d3b-88bf-cf75c92ec70c" src="https://user-images.githubusercontent.com/2040554/153887022-35002245-a6b3-4816-b593-bae570305fc9.png">

cleferman commented:
> @cleferman looks like the code lines and the numbers are aligning better now. Thanks!
> 
> What can we do to make the background of this container always the same color as the code area background? I can't seem to find the right variable for this.
> 
> <img alt="153886711-4eb9fe58-208a-4d3b-88bf-cf75c92ec70c" width="240" src="https://user-images.githubusercontent.com/2040554/153887022-35002245-a6b3-4816-b593-bae570305fc9.png">

@matthewbirtch I can use --center-channel-big. That's what we use for the copy background.

I just checked all the themes with that color and the indigo theme has issues:
![image](https://user-images.githubusercontent.com/14358537/154328253-ae580a3c-81d0-48f8-94f2-f65715c1b894.png)

The text is black. I noticed that for Onyx it works, the the text is white. Should I push it like this? Not entirely sure what to do abut the Indigo theme, but seems like it's a theme issue.

matthewbirtch commented:
> @matthewbirtch I can use --center-channel-big. That's what we use for the copy background.

I think the issue is that the code block background is using another color that is not coming from the main theme variables. I'm not sure how to leverage that color though

matthewbirtch commented:
> I just checked all the themes with that color and the indigo theme has issues: ![image](https://user-images.githubusercontent.com/14358537/154328253-ae580a3c-81d0-48f8-94f2-f65715c1b894.png)
> 
> The text is black. I noticed that for Onyx it works, the text is white. Should I push it like this? Not entirely sure what to do abut the Indigo theme, but seems like it's a theme issue.

Yes, I think another ticket was filed to resolve the indigo theme problem (https://mattermost.atlassian.net/browse/MM-41299)

cleferman commented:
> > @matthewbirtch I can use --center-channel-big. That's what we use for the copy background.
> 
> I think the issue is that the code block background is using another color that is not coming from the main theme variables. I'm not sure how to leverage that color though
> 
> <img alt="Screen Shot 2022-02-16 at 1 19 05 PM" width="284" src="https://user-images.githubusercontent.com/2040554/154330252-e3b97ea5-fcac-4c11-9ce9-c3587751b810.png">
> 
> h

I've put the background inside `.post-code .hljs` and it gets picked up because it's a stronger rule. Hope this is fine. Give it a go and let me know :).

matthewbirtch commented:
> I've put the background inside `.post-code .hljs` and it gets picked up because it's a stronger rule. Hope this is fine. Give it a go and let me know :).

I think it's ok from my standpoint. However, it looks like there might be a typo `--center-channel-big` should be `--center-channel-bg`

In addition, can we adjust the padding on `post-code__overlay` to `padding: 4px 4px 0px;`? Right now the padding is causing it to overlap with the border when the code is only 1 line.

As mentioned, we'll need to get the issue with the Indigo theme fixed asap (ideally before this is merged).

cleferman commented:
> > I've put the background inside `.post-code .hljs` and it gets picked up because it's a stronger rule. Hope this is fine. Give it a go and let me know :).
> 
> I think it's ok from my standpoint. However, it looks like there might be a typo `--center-channel-big` should be `--center-channel-bg`
> 
> In addition, can we adjust the padding on `post-code__overlay` to `padding: 4px 4px 0px;`? Right now the padding is causing it to overlap with the border when the code is only 1 line.
> 
> As mentioned, we'll need to get the issue with the Indigo theme fixed asap (ideally before this is merged).

Done.
I saw that the indigo issue is assigned to someone already.

matthewbirtch commented on sass/layout/_markdown.scss:
```diff
@@ -204,6 +217,7 @@ h6.markdown__heading {
         padding: 6.5px;
```
change to `padding: 8.5px`

This will prevent the hover actions from spanning outside the bounds of single line code blocks

matthewbirtch reviewed (approved):
2 minor tweaks and we're good to go

mattermod commented:
Triggering e2e testing with options:
```<nil>```

mattermod commented:
Successfully triggered e2e testing!
https://git.internal.mattermost.com/qa/cypress-ui-automation/-/pipelines/154878

matthewbirtch commented:
> * @matthewbirtch do we need to add `do not merge` label while we wait for Indigo theme issue to be fixed?

@jgilliam17 I think so, yes. Otherwise we'll have a problem with code display using that theme.

jgilliam17 commented:
@cleferman 
[E2E report ](https://automation-dashboard.vercel.app/cycles/0424977e-8f36-4852-a753-838c37cb9f83)has failures related to code theming. This may be related to background change requested here. Can you please take a look? Thanks 🙂 
<img width="854" alt="Screen Shot 2022-02-24 at 10 00 53 AM" src="https://user-images.githubusercontent.com/52937121/155549698-5594a752-9fe3-46d4-b390-da3f02342881.png">

hmhealey commented:
We can wait for Matt to respond, but I think a revert of that definitely makes sense since it's breaking the code themes. Using a dark theme with the default GitHub code theme makes it basically unreadable

<img width="586" alt="Screen Shot 2022-02-28 at 10 31 28 AM" src="https://user-images.githubusercontent.com/3277310/156010748-677dc67d-fbaa-4ba8-8227-2f4e0bccffe8.png">

matthewbirtch commented:
> @jgilliam17 @matthewbirtch not sure what to do about these e2e tests. They basically expect a different background color for each theme color, but that background color is now --center-channel-bg.
> 
> So I guess it's the issue you were talking about earlier in this thread @matthewbirtch. ![image](https://user-images.githubusercontent.com/14358537/155859630-6c1d7a41-e53a-4871-b29d-452c4144998b.png) The colors coming from the theme should actually overwrite the main color variables because if they don't there's no guarantee we'll actually see the text with all these combinations we have. Should we revert [this](https://github.com/mattermost/mattermost-webapp/pull/9655#issuecomment-1041982327)?

Interesting, I'm not really sure how the code themes are incorporated. I guess we can't really override the code themes based on the comment from @hmhealey. Are we able to somehow pull the code theme's background color for the background of the `post-code__overlay` container?

hmhealey commented:
I'm not actually sure about that. The way the code themes are applied is a bit of a black box. Unlike the regular theming where we have CSS variables, it's just a CSS file provided by the syntax highlighting library that we just download and apply to the whole page.

I suppose we could copy values from their CSS and add it as our own CSS variable. It would risk getting out of date, but that assumes highlight.js actually changes their style at any point

esethna commented:
@cleferman let us know if you have a clear path forward on this?

cleferman commented:
@esethna to be honest not really, I was waiting for a resolution on this. Meanwhile I reverted the change so the  background gets picked up from the theme.
Currently it looks like this (I guess it's similar to how it was before, the cypress tests linked here are passing):
![image](https://user-images.githubusercontent.com/14358537/157308570-ec0a41ee-c208-4edf-a58a-b80cf195a289.png)
I also added a fix for the copy container so it doesn't look like it's clipping when the code block is one liner:
![image](https://user-images.githubusercontent.com/14358537/157310405-41b984d2-f900-4c7b-9a5b-02e5b98b5c69.png)

esethna commented:
Thanks @cleferman. @matthewbirtch @hmhealey I'm not super familiar with all the context here, can we suggest a path forward? ie the copy button BG could use the center channel BG color, and then we don't put the code type in that container (as in the original design, see below)? Apologies, I might not have all the context of past conversations

![image](https://user-images.githubusercontent.com/14333569/157335733-6ba6399f-0c91-414a-9078-09a4b6775f80.png)

matthewbirtch commented:
> Thanks @cleferman. @matthewbirtch @hmhealey I'm not super familiar with all the context here, can we suggest a path forward? ie when hovering, the copy button BG could use the center channel BG color, and then we don't need to put the code type in that container (as in the original design, see below)? Apologies, I might not have all the context of past conversations
> 
> ![image](https://user-images.githubusercontent.com/14333569/157335733-6ba6399f-0c91-414a-9078-09a4b6775f80.png)

@esethna that is how we originally had it (using the center channel bg color), but the code themes are the problem. The background used for the code container is different from the center channel background and seems to be dictated by the code theme provider (e.g. github, monokai, etc). The problem was especially highlighted with the current Indigo theme which uses the github code theme right now which is a light theme (but there is a ticket out to fix this). Even if we fix that, we run the problem when custom themes are selected using different code themes.

Based on what @hmhealey has said, I think what @cleferman has done with the most recent commits is all we can do.

> I suppose we could copy values from their CSS and add it as our own CSS variable. It would risk getting out of date, but that assumes highlight.js actually changes their style at any point

@hmhealey do you think this is one possible way to get the colors right?

matthewbirtch commented:
> Thanks for the context. I'm not opposed to merging this as is (with maybe a few minor tweaks to padding etc via UX review), understanding that there is a UI flaw when there is a code theme that doesn't have the same BG color as the center channel. It's not the ideal outcome, but if we can't find a solution we're happy with then I wouldn't want to hold up the feature

Agreed! We could create a follow-up ticket to look in to improving this

jgilliam17 commented:
/update-branch

jgilliam17 commented:
Thanks @cleferman 
Re-tested  - the only remaining issue is the background color of the container. I can create a follow up to address that after merge. @matthewbirtch Is this acceptable?
@hmhealey let me know if we want to try your suggestion to fix using CSS here in this PR or in a follow up ticket?
Thanks 🙂

matthewbirtch commented:
> Re-tested - the only remaining issue is the background color of the container. I can create a follow up to address that after merge. @matthewbirtch Is this acceptable?

100% acceptable. Let's move forward with this and create a follow-up ticket to investigate

hmhealey commented:
> @hmhealey do you think this is one possible way to get the colors right?

Yeah, that'd be one way of doing it, but it's not great since they could get out of sync when we update highlight.js. There might be a better way of doing that though

> @hmhealey let me know if we want to try your suggestion to fix using CSS here in this PR or in a follow up ticket?

Yeah, let's do that separately. My proposal is a bit hacky, so I'd like if we spent some more time seeing if there's a better way to do it

jgilliam17 commented:
I'll update and merge after github issues are resolved. 
Ticket added for follow up https://mattermost.atlassian.net/browse/MM-42594

jgilliam17 reviewed (approved):
Sorry about the delay I missed recent replies
[E2E report](https://mattermost-cypress-report.s3.amazonaws.com/172033-8b10702-onprem-ent-webapp-pr-9655-mm-ee-test:8b10702/mochawesome.html) also looks good, no PR related failures

esethna commented:
@cleferman thanks for the PR! There was a bug found if you're keen on fixing it? https://mattermost.atlassian.net/browse/MM-42916

cleferman commented:
@esethna I'll see if I can do something about it over the weekend. Is there a github issue already?

amyblais commented:
There is a second bug ticket that @cleferman you may be interested to take: https://mattermost.atlassian.net/browse/MM-42777. cc @esethna

esethna commented:
@cleferman thanks! We don't have github issues for these but feel free to just link the JIRA tickets in the PR description

cleferman commented:
@amyblais I opened a PR for that here #10120.
@esethna Unfortunately I am not able to reproduce this. I was able to get my hands on a Macbook and tried reproducing it there as well on Safari 14.0.3, but no success. 
I noticed the context menu appears on a Mac if you hold ctrl while clicking on the trackpad, but I guess that's just the equivalent of a right click as I can't make a selection like that.

amyblais commented:
> Unfortunately I am not able to reproduce this. I was able to get my hands on a Macbook and tried reproducing it there as well on Safari 14.0.3, but no success. I noticed the context menu appears on a Mac if you hold ctrl while clicking on the trackpad, but I guess that's just the equivalent of a right click as I can't make a selection like that.

@cleferman A dev reproduced it and added some comments to the ticket https://mattermost.atlassian.net/browse/MM-42916. Feel free to reach out to Harrison Healey if you're having trouble reproducing.

cleferman commented:
@hmhealey the context menu feature is from the figma [mocks](https://www.figma.com/file/RV2tE26UUvoUlpx1mI4C32/MM-38770---Copy-Code-Blocks?node-id=1%3A47799). On the far left there's a screen called right-click menu.

I see it as well both on windows and on mac while long pressing without selecting anything. But not when selecting. From my perspective it's the right behaviour. (Maybe this is an issue on the newer version of Safari? I'm only able to test on 14.0.3) Check the gif below please. (The cursor turns yellow when I'm holding left-click)

![context-menu-behaviour](https://user-images.githubusercontent.com/14358537/161794169-6a568d09-8c19-4b02-aea6-5d1f64dfdd63.gif)

I've used the `ContextMenuTrigger` which was used for a different feature. But now I see that I've set `holdToDisplay={1000}`  instead of `-1` like in the other place where this is used. Again, this does seem to work as intended, but if it's still happening on Safari I guess we can set it to -1.

hmhealey commented:
Oh, I see it now. I didn't realize there was any new stuff on the left side of that design when I looked yesterday.

As for how it triggers, I think it's a distance thing. If you select text a few words and don't move your mouse very far, then it shows up, but if you move your mouse far enough before it would trigger, then it doesn't.

Either way though, I think we should just remove the `holdToDisplay` or set it to -1 would fix that for now unless we can find a way to make it only happen on a touch screen. I don't think it makes sense to include that behaviour when using a mouse or trackpad.

Let us know if you'd like to submit a PR for that. I think we're hoping to get that fixed in the next release, so we'd need something in in the next few days to get it for that. If not, no worries though.

cleferman commented:
I'll submit a PR in a few

esethna commented:
Thanks @cleferman !
